example_test.go

```diff
@@ -0,0 +1,106 @@
+package imbue_test
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/dogmatiq/imbue"
+)
+
+type Dependency1 string
+
+func (Dependency1) Close() error {
+	fmt.Println("closed dependency")
+	return nil
+}
+
+type Dependency2 string
+
+type SomeType struct {
+	Dep1         Dependency1
+	Dep2A, Dep2B Dependency2
+}
+
+type (
+	Red   imbue.IsName
+	Green imbue.IsName
+)
+
+func Example() {
+	con := imbue.New()
+
+	imbue.With3(
+		con,
+		func(
+			ctx *imbue.Context,
+			meaningless Dependency1,
+			red imbue.ByName[Red, Dependency2],
+			green imbue.ByName[Green, Dependency2],
+		) (SomeType, error) {
+			return SomeType{
+				Dep1:  meaningless,
+				Dep2A: red(),
+				Dep2B: green(),
+			}, nil
+		},
+	)
+
+	imbue.With0(
+		con,
+		func(
+			ctx *imbue.Context,
+		) (Dependency1, error) {
+			d := Dependency1("<dependency>")
+			ctx.Defer(d.Close)
+
+			return d, nil
+		},
+	)
+
+	imbue.With0(
+		con,
+		func(
+			ctx *imbue.Context,
+		) (Dependency2, error) {
+			return Dependency2("<red>"), nil
+		},
+		imbue.Name[Red](),
```

Comment:
You'd need to have more type information in the signatures like:

```go
func With(
	ctx context.Context,
) imbue.Named[Green, Dependency1] {
	return imbue.WithName(Green, Dependency1)
}
```

```go
InvokeWith1(
		ctx context.Context,
		func(
			ctx context.Context,
			dep imbue.FromName2[Red, Green, Dependency1],			
		) error {
// ....
```

Reply:
I thought that `imbue.WithName()` will store the `Dependency1` in its internal name-to-type map  and would still return `Dependency1` to avoid complex types.  Eventually `imbue.FromName()` would bring the named variable inside `imbue.InvokeWithN`, but that would be too implicit for the user.

Reply:
> I thought that imbue.WithName() will store the Dependency1 in its internal name-to-type map

Yeah, I didn't consider that approach, that would work, but I still don't know how it would work when you need to request something by name. I don't think it's too implicit, I just don't know how to implement it.

Reply:
What I was trying to think is really cumbersome and inapplicable to the design of `imbue` package.